### PR TITLE
Fix array data copy behavior for numpy 1 and 2 consistency

### PR DESCRIPTION
**Pull request by andy-sweet**
# References and relevant issues

Failure in our prerelease tests which picked up `numpy==2.0.0rc2`: https://github.com/napari/napari/issues/6871#issue-2263425683

Related discussion on Zulip: https://napari.zulipchat.com/#narrow/stream/212875-general/topic/handling.20the.20numpy.202.2E0.20release

# Description
This fixes the coercion of data using the `Array` pydantic custom class so that it works for both numpy 1 and 2, and only makes a copy if needed.

Instead we could consider using `np.asarray`, but we'll also need to replace the `ndmin` argument (which `np.asarray` does not have), by using `np.expand_dims` or similar. I have done that it in one other place where I don't need those fancy arguments.

### CODE REVIEW

File: napari/utils/events/custom_types.py [hunk 1]
```diff
@@ -35,7 +35,16 @@ def validate_type(cls, val):
...
 
-        result = np.array(val, dtype=dtype, copy=False, ndmin=len(shape))
+        # In numpy 1, copy=False means that copy is only made if needed.
+        # In numpy 2, copy=False errors if a copy is needed, and instead
+        # the equivalent is copy=None.
+        # We would normally use np.asarray instead, but here we need the
+        # behavior associated with ndmin.
+        # https://numpy.org/devdocs/numpy_2_0_migration_guide.html#adapting-to-changes-in-the-copy-keyword
+        copy = None
+        if np.lib.NumpyVersion(np.__version__) < '2.0.0b1':
```
Comment (andy-sweet) [inline napari/utils/events/custom_types.py]: There's probably a better comparison than this, but I'm sleepy and this is the best I've got right now.

**Comment by andy-sweet** [comment]
There are other uses of `np.array(..., copy=False, ...)` in our code. I think we could replace most of those with `np.asarray` (or as here if needed) as I doubt we ever want to error if a copy must be made.

**Review by jni — APPROVED** [review]

**Comment by jni** [comment]
@andy-sweet are you planning on changing those in this PR?

**Comment by andy-sweet** [comment]
> @andy-sweet are you planning on changing those in this PR?

I'm out for the next 24 hours, but should be able do that soon after. May as well get it done in this PR, so yes!

**Comment by andy-sweet** [comment]
My process here was to grep/search for `copy=False` and find usage with `np.array` and similar.

The following are some similar usages that I did not change.

- [`np.nan_to_num` used in `layer_utils.py`](https://github.com/napari/napari/blob/07dbc0a8f49cc2c73f3215fb75988e1c1bfcf1ab/napari/layers/utils/layer_utils.py#L361): no change to its `copy` arg [is documented](https://numpy.org/devdocs/reference/generated/numpy.nan_to_num.html)
- [`np.md.array` used in `colors.py` (vendored from matplotlib)](https://github.com/napari/napari/blob/07dbc0a8f49cc2c73f3215fb75988e1c1bfcf1ab/napari/utils/colormaps/vendored/colors.py#L872): no change to its `copy` arg [is documented](https://numpy.org/devdocs/reference/generated/numpy.ma.array.html#numpy-ma-array)
- [`np.meshgrid` used in `points.py`](https://github.com/napari/napari/blob/07dbc0a8f49cc2c73f3215fb75988e1c1bfcf1ab/napari/layers/points/points.py#L2352): no change to its `copy` arg [is documented](https://numpy.org/devdocs/reference/generated/numpy.meshgrid.html#numpy-meshgrid)

There is also some usage of `np.array(..., copy=False, ...)` in [`colors.py` vendored from matplotlib](https://github.com/napari/napari/blob/07dbc0a8f49cc2c73f3215fb75988e1c1bfcf1ab/napari/utils/colormaps/vendored/colors.py#L1243). I did not change these because they have [not been changed upstream in matplotlib](https://github.com/matplotlib/matplotlib/blob/2834da176a925e6ae8e8e895b402f1f456fcf057/lib/matplotlib/colors.py#L2205). If these cause an issue, we could just vendor their changes or re-vendor the whole module.

**Comment by andy-sweet** [comment]
Pending CI checks, this should be ready now. I pulled out the numpy version check into the module import, so that it's only executed once. This definition could also be moved to somewhere like `napari.utils.misc` if needed elsewhere.

**Comment by andy-sweet** [comment]
Note that I don't expect this to fix the pre release tests. There are at least some vispy issues lurking as mentioned on Zulip: https://napari.zulipchat.com/#narrow/stream/212875-general/topic/handling.20the.20numpy.202.2E0.20release/near/440434390

**Pull request merged**